From b117b5b6605f9f7588b5b7200b66081a71fb3083 Mon Sep 17 00:00:00 2001
From: Josef Schlehofer <pepe.schlehofer@gmail.com>
Date: Tue, 10 Aug 2021 11:45:41 +0200
Subject: [PATCH] treewide: add support for 80+80 MHz channel

OpenWrt's iwinfo exposes many values like VHT80_80 [1] and HE80_80 [2],
but while testing miniPCIe card Compex WLE1216V5 [3] inserted to Turris
Omnia router, I noticed that in LuCI, it is possible to configure
channel width '20 MHz', '40 MHz', '80 MHz' and '160 MHz', but according
to the datasheet [4], this miniPCie card supports 80+80 MHz and not 160 MHz, which is different.
It avoid DFS channels and operate on two channels simultanously.

Before this commit:
- LuCI configures this card by looking at uci as "htmode": "VHT160"

After this commit:
- LuCI

-- commit message needs to be improved --

[1] https://git.openwrt.org/?p=project/iwinfo.git;a=blob;f=include/iwinfo.h;h=8469ee70f5e36342780817236d3994a440e0788e;hb=HEAD#l91
[2] https://git.openwrt.org/?p=project/iwinfo.git;a=blob;f=include/iwinfo.h;h=8469ee70f5e36342780817236d3994a440e0788e;hb=HEAD#l97
[3] Network controller [0280]: Qualcomm Atheros QCA9984 802.11ac Wave 2 Wireless Network Adapter [168c:0046]
[4] https://compex.com.sg/wp-content/uploads/2021/01/wle1216v5-20-v1.3-sl-227.pdf

Signed-off-by: Josef Schlehofer <pepe.schlehofer@gmail.com>
---
 libs/rpcd-mod-luci/src/luci.c                               | 3 +++
 modules/luci-base/htdocs/luci-static/resources/network.js   | 2 ++
 modules/luci-compat/luasrc/view/cbi/wireless_modefreq.htm   | 1 +
 .../htdocs/luci-static/resources/view/network/wireless.js   | 6 ++++--
 4 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/libs/rpcd-mod-luci/src/luci.c b/libs/rpcd-mod-luci/src/luci.c
index ff656ec212..3bc7ad0162 100644
--- a/libs/rpcd-mod-luci/src/luci.c
+++ b/libs/rpcd-mod-luci/src/luci.c
@@ -989,6 +989,9 @@ static bool rpc_luci_get_iwinfo(struct blob_buf *buf, const char *devname,
 		if (nret & IWINFO_HTMODE_HE80)
 			blobmsg_add_string(buf, NULL, "HE80");
 
+		if (nret & IWINFO_HTMODE_HE80_80)
+			blobmsg_add_string(buf, NULL, "HE80+80");
+
 		if (nret & IWINFO_HTMODE_HE160)
 			blobmsg_add_string(buf, NULL, "HE160");
 
diff --git a/modules/luci-base/htdocs/luci-static/resources/network.js b/modules/luci-base/htdocs/luci-static/resources/network.js
index 8c9ee255ff..f895da3fe8 100644
--- a/modules/luci-base/htdocs/luci-static/resources/network.js
+++ b/modules/luci-base/htdocs/luci-static/resources/network.js
@@ -3376,10 +3376,12 @@ WifiDevice = baseclass.extend(/** @lends LuCI.network.WifiDevice.prototype */ {
 	 *  - `VHT20` - applicable to IEEE 802.11ac, 20 MHz wide channels
 	 *  - `VHT40` - applicable to IEEE 802.11ac, 40 MHz wide channels
 	 *  - `VHT80` - applicable to IEEE 802.11ac, 80 MHz wide channels
+	 *  - `VHT80_80` - applicable to IEEE 802.11ac, 80+80 MHz wide channels
 	 *  - `VHT160` - applicable to IEEE 802.11ac, 160 MHz wide channels
 	 *  - `HE20` - applicable to IEEE 802.11ax, 20 MHz wide channels
 	 *  - `HE40` - applicable to IEEE 802.11ax, 40 MHz wide channels
 	 *  - `HE80` - applicable to IEEE 802.11ax, 80 MHz wide channels
+	 *  - `HE80_80` - applicable to IEEE 802.11ax, 80+80 MHz wide channels
 	 *  - `HE160` - applicable to IEEE 802.11ax, 160 MHz wide channels
 	 */
 	getHTModes: function() {
diff --git a/modules/luci-compat/luasrc/view/cbi/wireless_modefreq.htm b/modules/luci-compat/luasrc/view/cbi/wireless_modefreq.htm
index eeb1d5c5cb..fb9ab9a61a 100644
--- a/modules/luci-compat/luasrc/view/cbi/wireless_modefreq.htm
+++ b/modules/luci-compat/luasrc/view/cbi/wireless_modefreq.htm
@@ -44,6 +44,7 @@
 			'VHT20', '20 MHz', htmodes.VHT20,
 			'VHT40', '40 MHz', htmodes.VHT40,
 			'VHT80', '80 MHz', htmodes.VHT80,
+			'VHT80_80', '80+80 MHz', htmodes.VHT80_80,
 			'VHT160', '160 MHz', htmodes.VHT160
 		]
 	};
diff --git a/modules/luci-mod-network/htdocs/luci-static/resources/view/network/wireless.js b/modules/luci-mod-network/htdocs/luci-static/resources/view/network/wireless.js
index f94640ee7d..5d5f31b053 100644
--- a/modules/luci-mod-network/htdocs/luci-static/resources/view/network/wireless.js
+++ b/modules/luci-mod-network/htdocs/luci-static/resources/view/network/wireless.js
@@ -362,12 +362,14 @@ var CBIWifiFrequencyValue = form.Value.extend({
 					'VHT20', '20 MHz', htmodelist.VHT20,
 					'VHT40', '40 MHz', htmodelist.VHT40,
 					'VHT80', '80 MHz', htmodelist.VHT80,
+					'VHT80_80', '80+80 MHz', htmodelist.VHT80_80,
 					'VHT160', '160 MHz', htmodelist.VHT160
 				],
 				'ax': [
 					'HE20', '20 MHz', htmodelist.HE20,
 					'HE40', '40 MHz', htmodelist.HE40,
 					'HE80', '80 MHz', htmodelist.HE80,
+					'HE80_80', '80+80 MHz', htmodelist.HE80_80,
 					'HE160', '160 MHz', htmodelist.HE160
 				]
 			};
@@ -452,9 +454,9 @@ var CBIWifiFrequencyValue = form.Value.extend({
 
 		this.setValues(mode, this.modes);
 
-		if (/HE20|HE40|HE80|HE160/.test(htval))
+		if (/HE20|HE40|HE80|HE80_80|HE160/.test(htval))
 			mode.value = 'ax';
-		else if (/VHT20|VHT40|VHT80|VHT160/.test(htval))
+		else if (/VHT20|VHT40|VHT80|VHT80_80|VHT160/.test(htval))
 			mode.value = 'ac';
 		else if (/HT20|HT40/.test(htval))
 			mode.value = 'n';
-- 
2.32.0

